File: ros2/workspace/src/simple_ros_cpp/src/move_turtle.cpp
#include <iostream>
#include <chrono> 
#include "rclcpp/rclcpp.hpp"
#include "turtlesim/msg/color.hpp"
#include "turtlesim/msg/pose.hpp"
#include "geometry_msgs/msg/twist.hpp"

using namespace std; 
void printHello(); 

class MoveTurtle : public rclcpp::Node {
    public:
        MoveTurtle() : Node("move_turtle") {
            auto qos_profile = rclcpp::QoS(rclcpp::KeepLast(10)).reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE)
                .durability(RMW_QOS_POLICY_DURABILITY_VOLATILE); 
            _pose_sub = create_subscription<turtlesim::msg::Pose>(
                "turtle1/pose", qos_profile, 
                std::bind(&MoveTurtle::pose_sub_callback, this, std::placeholders::_1)); 
            _color_sub = create_subscription<turtlesim::msg::Color>(
                "turtle1/color", qos_profile,  
                std::bind(&MoveTurtle::color_sub_callback, this, std::placeholders::_1)); 

            _pub = create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", qos_profile);
            _twist_pub_timer = create_wall_timer(1s, std::bind(&MoveTurtle::twist_pub, this)); 
            _update_timer = create_wall_timer(17ms, std::bind(&MoveTurtle::update, this)); 
        }
    private:
        int _count; 
        turtlesim::msg::Color _color; 
        turtlesim::msg::Pose _pose; 
        geometry_msgs::msg::Twist _twist; 

        rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr _pose_sub; 
        rclcpp::Subscription<turtlesim::msg::Color>::SharedPtr _color_sub; 
        rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr _pub; 
        rclcpp::TimerBase::SharedPtr _twist_pub_timer; 
        rclcpp::TimerBase::SharedPtr _update_timer; 
        
        void pose_sub_callback(const turtlesim::msg::Pose::SharedPtr msg) {
            _pose = *msg; 
        }
        void color_sub_callback(const turtlesim::msg::Color::SharedPtr msg) {
            _color = *msg; 
        }

        void twist_pub() {
            _pub->publish(_twist); 
        }

        void update() {
            _twist.linear.x += 0.004; 
            _twist.angular.z = 1.0; 
        }
}; 

int main()
{
    rclcpp::init(0, nullptr);
    auto node = std::make_shared<MoveTurtle>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2/workspace/src/SimplePkg/SimplePkg/move_turtle.py
import numpy as np
import os
import rclpy 
from rclpy.node import Node
from std_msgs.msg import String 
from rclpy.qos import QoSProfile 
from rclpy.qos import QoSHistoryPolicy, QoSReliabilityPolicy, QoSDurabilityPolicy
from rclpy.qos import qos_profile_sensor_data as qos
from turtlesim.msg import Color, Pose, Twist

class MoveTurtle(Node):
    def __init__(self):
        super().__init__('move_turtle')
        self.qos_profile = QoSProfile(history=QoSHistoryPolicy.KEEP_LAST, 
                                      reliability=QoSReliabilityPolicy.BEST_EFFORT, 
                                      durability=QoSDurabilityPolicy.VOLATILE,
                                      depth=10)   
        
        self.create_timer(0.1, self.twist_pub)
        self.create_timer(1/60, self.update)

        self.massage_publisher = self.create_publisher(Twist, 'turtle1/cmd_vel', self.qos_profile)
        self.create_subscription(Pose, "turtle1/pose", self.pose_callback)
        self.create_subscription(Color, "turtle1/color_sensor", self.color_callback)

        self.count = 0
        self.twist = Twist()
        self.pose = Pose()
        self.color = Color()

    def twist_pub(self):
        self.pub.publish(self.twist)

    def pose_callback(self, msg:Pose):
        self.pose = msg

    def color_callback(self, msg:Color):
        self.color = msg

    def print_hello(self):
        self.get_logger().info(f"print_hello message..")
        self.count += 1

    def update(self):
        self.twist.linear.x += 0.001
        self.twist.angular.z = 1.0

def main(args=None):
    rclpy.init(args=args)
    node = MoveTurtle()
    try :
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("Keyboard interrupted!!")
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    print('Hello ROS2 Simple Package second')
    main()

    